File: jamcatch/engine/src/main/java/org/example/javarca/engine/Renderer.java
package org.example.javarca.engine;

import java.util.ServiceLoader;

/**
 * Interface to be implemented on top of a library or framework that renders a running game.
 * The implementation should use {@link GameLoop} and {@link GameState},
 */
public interface Renderer {
    static void launch() {
        System.out.println("Welcome to Javarcade!");

        Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
            System.out.println("Unexpected crash!" + throwable.getMessage());
            throwable.printStackTrace();
            throw new RuntimeException(throwable);
        });

        var impl = ServiceLoader.load(Renderer.class).findFirst();
        impl.ifPresentOrElse(Renderer::run, () -> System.out.println("No renderer implementation available"));
    }

    void run();
}

File: jamcatch/engine/src/jvmMain/kotlin/org/example/javarca/engine/Renderer.kt
package org.example.javarca.engine

import java.util.*
import java.util.function.Consumer

/**
 * Interface to be implemented on top of a library or framework that renders a running game. The
 * implementation should use [GameLoop] and [GameState],
 */
interface Renderer {
    fun run()

    companion object {
        fun launch() {
            println("Welcome to Javarcade!")

            Thread.setDefaultUncaughtExceptionHandler { thread: Thread, throwable: Throwable ->
                println("Unexpected crash!" + throwable.message)
                throwable.printStackTrace()
                throw RuntimeException(throwable)
            }

            val impl = ServiceLoader.load(Renderer::class.java).findFirst()
            impl.ifPresentOrElse(Consumer { obj: Renderer? -> obj!!.run() }) {
                println("No renderer implementation available")
            }
        }
    }
}
